pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -363,13 +401,12 @@ void MultiTopicsConsumerImpl::closeAsync(ResultCallback callback) {
 
     auto self = shared_from_this();
     int numConsumers = 0;
-    consumers_.forEach(
-        [&numConsumers, &self, callback](const std::string& name, const ConsumerImplPtr& consumer) {
-            numConsumers++;
-            consumer->closeAsync([self, name, callback](Result result) {
-                self->handleSingleConsumerClose(result, name, callback);
-            });
+    for (const auto& item : consumers_.toPairVector()) {
```

Comment:
The effect of these traversal methods looks the same. BTW, `toPairVector` is marked with 'only used for tests'

Reply:
> The effect of these traversal methods looks the same. 
This method is to copy a copy to traverse. If this is not done, a pointer exception will occur.

> toPairVector is marked with 'only used for tests'.

Maybe need to find another better way, I will look at it again.

Reply:
There is a `forEachValue` method that I've provided in `SynchronizedHashMap`.

```c++
consumers_.forEachValue([&numConsumers/* other captures... */](const ConsumerImplPtr& consumer) {
    numConsumers++;
    consumer->closeAsync(/* ... */);
});
```

Reply:
@RobertIndie @BewareMyPower The current implementation removes or cleans up the `consumers_` in the `handSingleConsumerClose` method, so there is a risk that the nullptr.

https://github.com/apache/pulsar/blob/7f6972766f1f8a989f477cbb3085d0660be53bd8/pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc#L369-L375

https://github.com/apache/pulsar/blob/7f6972766f1f8a989f477cbb3085d0660be53bd8/pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc#L392

https://github.com/apache/pulsar/blob/7f6972766f1f8a989f477cbb3085d0660be53bd8/pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc#L409

So use `toPairVector` to re-copy a collection for traversal.  Is there a better way?

Reply:
> So there is a risk that the nullptr.

At the very beginning, there is no chance that a `std::shared_ptr` becomes null unless the `reset()` method is explicitly called. You might mean the `std::shared_ptr` points to an invalid memory area.

However, it's still impossible. First, the `forEachValue` method is protected by the internal lock, so the function argument accepts a non-null `consumer`. Second, even if the value was removed from the map when the callback of `closeAsync` method (`handleSingleConsumerClosed`) was executed, the object that the `ConsumerImplPtr` object points would still be valid. Third, the increment and decrement of the reference count of a `std::shared_ptr` is thread safe.

The key point is that `ConsumerImplPtr` is a `std::shared_ptr`, not a trivial reference. The lifetime ends only when the reference count become 0. For example,

```c++
#include <iostream>
#include <memory>
#include <vector>
using namespace std;

struct Object {
    Object() { cout << "ctor" << endl; }
    ~Object() { cout << "dtor" << endl; }
    void f() { cout << (void*)this << endl; }
};

int main(int argc, char* argv[]) {
    using ObjectPtr = std::shared_ptr<Object>;
    vector<ObjectPtr> objects;
    objects.emplace_back(std::make_shared<Object>());
    auto object = objects[0];
    cout << object.use_count() << endl;  // 2
    objects.clear();
    cout << object.use_count() << endl;  // 1
    object->f();                         // it works well because the reference count is still 1
    return 0;
}
```

Reply:
@BewareMyPower Sorry, maybe I didn't make it clear. The problem now is that we did a `remove` operation in the function of `forEachValue`.

case this:
```c++
class TestClass{
   public:
    std::string key;
    void print(){
        std::cout << "test ttt." << std::endl;
    };
};

TEST(SchemaTest, ttt) {
    SynchronizedHashMap<std::string, std::shared_ptr<TestClass>> map;

    std::shared_ptr<TestClass> test1 = std::make_shared<TestClass>();
    test1->key = "test1";
    std::shared_ptr<TestClass> test2 = std::make_shared<TestClass>();
    test2->key = "test2";
    map.emplace(test1->key, test1);
    map.emplace(test2->key, test2);

    map.forEachValue([&map](std::shared_ptr<TestClass> test) {
        test->print();
        map.remove(test->key);
    });
}

```

will result:

```shell
/Users/baozi/workspace/github/pulsar-cpp/pulsar/pulsar-client-cpp/tests/main --gtest_filter=SchemaTest.ttt:SchemaTest/*.ttt:SchemaTest.ttt/*:*/SchemaTest.ttt/*:*/SchemaTest/*.ttt --gtest_color=no
Testing started at 18:0009 ...
test:test2
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

```

Reply:
I got your point now. However, your test is not exactly the same case. A C++ container (`vector`, `map`, etc.) should not be modified in a range for loop. For example,

```c+
    std::map<int, int> m;
    m[1] = 0;
    m[2] = 0;
    for (const auto& kv : m) {
        m.erase(kv.first);
    }
```

It's because a range for loop just iterates over the underlying RB-tree (or hash table against a `std::unordered_map`). Modifying it can make the next iterator invalid.

----

However, the case here is different that you call `map.remove` directly in the range for loop, while the `handleSingleConsumerClose` is a callback of `closeAsync` that is called in another thread. For example, here is a test I added for the close of multi-topics consumer (still use `forEach`, and change the log of `handleSingleConsumerClose` to info):

```c++
TEST(ConsumerTest, testMultiTopicsConsumerClose) {
    Client client(lookupUrl);

    Consumer consumer;
    client.subscribe({"tp1", "tp2", "tp3"}, "sub", consumer);

    auto send = [&client](const std::string& topic) {
        Producer producer;
        client.createProducer(topic, producer);
        producer.send(MessageBuilder().setContent("hello").build());
    };
    send("tp1");
    send("tp2");
    send("tp3");

    while (true) {
        Message msg;
        auto result = consumer.receive(msg, 1000);
        if (result != ResultOk) {
            break;
        }
        LOG_INFO("Received " << msg.getDataAsString() << " from " << msg.getTopicName() << " "
                             << msg.getMessageId());
    }

    auto result = consumer.close();
    LOG_INFO("Consumer close: " << result);
}
```

The outputs are:

```
2022-08-23 13:22:10.567 INFO  [0x11d299600] ConsumerImpl:991 | [persistent://public/default/tp3, sub, 2] Closing consumer for topic persistent://public/default/tp3
2022-08-23 13:22:10.567 INFO  [0x11d299600] ConsumerImpl:991 | [persistent://public/default/tp2, sub, 1] Closing consumer for topic persistent://public/default/tp2
2022-08-23 13:22:10.567 INFO  [0x11d299600] ConsumerImpl:991 | [persistent://public/default/tp1, sub, 0] Closing consumer for topic persistent://public/default/tp1
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] ConsumerImpl:1042 | [persistent://public/default/tp3, sub, 2] Closed consumer 2
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] MultiTopicsConsumerImpl:395 | Closing the consumer for partition - persistent://public/default/tp3 numberTopicPartitions_ - 3
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] ConsumerImpl:1042 | [persistent://public/default/tp2, sub, 1] Closed consumer 1
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] MultiTopicsConsumerImpl:395 | Closing the consumer for partition - persistent://public/default/tp2 numberTopicPartitions_ - 2
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] ConsumerImpl:1042 | [persistent://public/default/tp1, sub, 0] Closed consumer 0
2022-08-23 13:22:10.572 INFO  [0x70000d14e000] MultiTopicsConsumerImpl:395 | Closing the consumer for partition - persistent://public/default/tp1 numberTopicPartitions_ - 1
```

The id of the thread that calls `closeAsync` (which calls `forEach` on a `SynchronizedHashMap`) is 0x11d299600, while the id of the thread that calls `handleSingleConsumerClose` (which calls `remove` or `clear` on a `SynchronizedHashMap`) is 0x70000d14e000.

**BTW, I found there is still a chance that `handleSingleConsumerClose` is called directly in the current thread. I think it's a bug that needs to be fixed. I might open another PR to fix it.**

Reply:
Get it, thanks.

Reply:
Wait for https://github.com/apache/pulsar/pull/17239 to fix the problem.

Reply:
Rebased #17239.